Propagate PO line changes to corresponding SO lines

adi-stefan opened the pull request.

adi-stefan committed:
Propagate PO line changes to corresponding SO lines

adi-stefan committed:
Merge branch 'new_dawn_uat' into new_dawn_uat_PO_ProjectId_To_SO

teosarca commented on backend/de.metas.business/src/main/java/de/metas/order/model/interceptor/C_Order_Project.java:
```diff
@@ -50,34 +58,61 @@
 
 	@DocValidate(timings = { ModelValidator.TIMING_BEFORE_COMPLETE })
 	public void po_populateProjectIfNeeded(@NonNull final I_C_Order order)
+	{
```
in this case, i think `po_populateProjectIfNeeded` could be renamed to `beforeComplete`,
since we have populateProjectIfNeeded

adi-stefan replied:
done

teosarca commented on backend/de.metas.purchasecandidate.base/src/main/java/de/metas/purchasecandidate/PurchaseCandidateRepository.java:
```diff
@@ -707,6 +711,19 @@ public List<PurchaseCandidate> getAllByPurchaseOrderId(@NonNull final OrderId pu
+				.stream()
+				.map(I_C_PurchaseCandidate_Alloc::getC_PurchaseCandidate_ID)
+				.map(PurchaseCandidateId::ofRepoId)
+				.map(this::getById)
```
SQL N+1

adi-stefan replied:
fixed.

teosarca commented on backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/purchaseOrder/purchaseOrderProjectToSO.feature:
```diff
@@ -0,0 +1,110 @@
+  @from:cucumber
+  Scenario:
+    Given metasfresh contains M_Products:
+      | Identifier | OPT.IsSold | OPT.IsPurchased |
```
pls avoid OPT.

adi-stefan replied:
done

teosarca commented on backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/purchaseOrder/purchaseOrderProjectToSO.feature:
```diff
@@ -0,0 +1,110 @@
+      | Identifier |
+      | ps_1       |
+    And metasfresh contains M_PriceLists
+      | Identifier | M_PricingSystem_ID.Identifier | OPT.C_Country.CountryCode | C_Currency.ISO_Code | SOTrx | IsTaxIncluded | PricePrecision |
```
pls avoid .Identigier and OPT.

teosarca replied:
u can drop IsTaxIncluded and PricePrecision  becuase not relevant for ur test

adi-stefan replied:
done

teosarca commented on backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/purchaseOrder/purchaseOrderProjectToSO.feature:
```diff
@@ -0,0 +1,110 @@
+      | pl_1       | ps_1                          | DE                        | EUR                 | true  | false         | 2              |
+      | pl_2       | ps_1                          | DE                        | EUR                 | false | false         | 2              |
+    And metasfresh contains M_PriceList_Versions
+      | Identifier | M_PriceList_ID.Identifier | ValidFrom  |
```
avoid .Identiier
drop ValidFrom

adi-stefan replied:
dropped

teosarca commented on backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/purchaseOrder/purchaseOrderProjectToSO.feature:
```diff
@@ -0,0 +1,110 @@
+      | plv_1      | pl_1                      | 2021-04-01 |
+      | plv_2      | pl_2                      | 2021-04-01 |
+    And metasfresh contains M_ProductPrices
+      | Identifier | M_PriceList_Version_ID.Identifier | M_Product_ID.Identifier | PriceStd | C_UOM_ID.X12DE355 | C_TaxCategory_ID.InternalName |
```
* drop | Identifier | column
* don't use .Identifier suffixes
* u can use C_UOM_ID instead of C_UOM_ID.X12DE355
* drop C_TaxCategory_ID.InternalName

adi-stefan replied:
done.

teosarca commented on backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/purchaseOrder/purchaseOrderProjectToSO.feature:
```diff
@@ -0,0 +1,110 @@
+      | pp_1       | plv_1                             | p_1                     | 10.0     | PCE               | Normal                        |
+      | pp_2       | plv_2                             | p_1                     | 10.0     | PCE               | Normal                        |
+
+    And metasfresh contains M_DiscountSchemas:
```
why do we need discount schemas?

adi-stefan replied:
because the lookup logic for vendors fails if no discount schema is defined. See: de.metas.purchasecandidate.VendorProductInfoService#getVendorProductInfos

teosarca commented on backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/purchaseOrder/purchaseOrderProjectToSO.feature:
```diff
@@ -0,0 +1,110 @@
+      | endvendor_1              | p_1                     |
+
+    And metasfresh contains C_Orders:
+      | Identifier | IsSOTrx | C_BPartner_ID.Identifier | DateOrdered | OPT.PreparationDate  |
```
avoid .Identifier suffixes and OPT. prefixes

adi-stefan replied:
done

teosarca commented on backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/purchaseOrder/purchaseOrderProjectToSO.feature:
```diff
@@ -0,0 +1,110 @@
+    And metasfresh contains C_Orders:
+      | Identifier | IsSOTrx | C_BPartner_ID.Identifier | DateOrdered | OPT.PreparationDate  |
+      | o_1        | true    | endcustomer_1            | 2021-04-17  | 2021-04-16T21:00:00Z |
+    And metasfresh contains C_OrderLines:
```
avoid .Identifier suffixes and OPT. prefixes

teosarca replied:
have multiple sales order lines

teosarca commented on backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/purchaseOrder/purchaseOrderProjectToSO.feature:
```diff
@@ -0,0 +1,110 @@
+      | C_OrderLine_ID.Identifier | C_Order_ID.Identifier | OPT.DateOrdered | M_Product_ID.Identifier | QtyOrdered | qtydelivered | qtyinvoiced | price | discount | currencyCode | processed | C_Project_ID |
+      | pol_1                     | po_1                  | 2021-04-16      | p_1                     | 10         | 0            | 0           | 10    | 0        | EUR          | true      | proj1        |
+
+## SO lines populated
```
IMHO this validation shall go after the sales order is completed

adi-stefan replied:
no, this is only populated once the resulting PO is completed.

teosarca commented on backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/purchaseOrder/purchaseOrderProjectToSO.feature:
```diff
@@ -0,0 +1,110 @@
+    And metasfresh contains C_BPartners without locations:
+      | Identifier    | OPT.IsVendor | OPT.IsCustomer | M_PricingSystem_ID.Identifier | OPT.PO_DiscountSchema_ID.Identifier |
+      | endcustomer_1 | N            | Y              | ps_1                          |                                     |
+      | endvendor_1   | Y            | N              | ps_1                          | ds_1                                |
```
endvendor :)))

teosarca commented on backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/order/C_OrderLine_StepDef.java:
```diff
@@ -661,7 +667,23 @@ private void validateOrderLine(@NonNull final I_C_OrderLine orderLine, @NonNull
+			}
+			else if (orderLine.getC_Project_ID() > 0)
+			{
+				project = load(orderLine.getC_Project_ID(), I_C_Project.class);
```
pls use de.metas.project.service.ProjectRepository#getById

adi-stefan replied:
done

teosarca commented on backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/order/C_Order_StepDef.java:
```diff
@@ -835,6 +839,22 @@ private void validateOrder(@NonNull final DataTableRow row)
+			}
+			else if (order.getC_Project_ID() > 0)
+			{
+				project = load(order.getC_Project_ID(), I_C_Project.class);
```
pls use de.metas.project.service.ProjectRepository#getById

adi-stefan replied:
done

teosarca commented on backend/de.metas.business/src/main/java/de/metas/order/IPOLineProjectPropagator.java:
```diff
@@ -0,0 +1,39 @@
+ * Its main usage is to ensure a project set to a purchase order line is propagated to any corresponding sales order line (via purchase candidate).
+ */
+@FunctionalInterface
+public interface IPOLineProjectPropagator
```
PurchaseOrderProjectListener

teosarca commented on backend/de.metas.business/src/main/java/de/metas/order/IPOLineProjectPropagator.java:
```diff
@@ -0,0 +1,39 @@
+@FunctionalInterface
+public interface IPOLineProjectPropagator
+{
+	void propagateProjectId(@NonNull ProjectId projectId, @NonNull Collection<OrderAndLineId> purchaseOrderLineIds);
```
onCreated

teosarca commented on backend/de.metas.business/src/main/java/de/metas/order/PurchaseOrderProjectListener.java:
```diff
@@ -0,0 +1,39 @@
+ * <p>
+ * Its main usage is to ensure a project set to a purchase order line is propagated to any corresponding sales order line (via purchase candidate).
+ */
+@FunctionalInterface
```
drop functional annotation

teosarca commented on backend/de.metas.business/src/main/java/de/metas/order/PurchaseOrderProjectListener.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Collection;
+
+/**
+ * Represents a functional interface for performing necessary propagation of a project ID to a collection of order line IDs.
```
fix javadoc

adi-stefan replied:
done.

adi-stefan committed:
QA

adi-stefan committed:
QA

adi-stefan committed:
Merge branch 'new_dawn_uat' into new_dawn_uat_PO_ProjectId_To_SO

adi-stefan committed:
Merge branch 'new_dawn_uat' into new_dawn_uat_PO_ProjectId_To_SO

adi-stefan committed:
cleanup

adi-stefan committed:
added scenario name

adi-stefan committed:
Merge branch 'new_dawn_uat' into new_dawn_uat_PO_ProjectId_To_SO

adi-stefan committed:
increase timeout to allow backend to catch up

adi-stefan committed:
Merge remote-tracking branch 'origin/new_dawn_uat_PO_ProjectId_To_SO' into new_dawn_uat_PO_ProjectId_To_SO

adi-stefan committed:
Merge branch 'new_dawn_uat' into new_dawn_uat_PO_ProjectId_To_SO

adi-stefan committed:
move to different executor

adi-stefan committed:
Merge remote-tracking branch 'origin/new_dawn_uat_PO_ProjectId_To_SO' into new_dawn_uat_PO_ProjectId_To_SO

adi-stefan committed:
Cleanup/simplify failing cucumber

adi-stefan committed:
Merge branch 'new_dawn_uat' into new_dawn_uat_PO_ProjectId_To_SO

adi-stefan committed:
Cleanup/simplify failing cucumber

teosarca committed:
Simplify purchaseSimulation.feature data structure by removing redundant columns in PriceLists, Product Prices, and BPartner_Locations tables.

teosarca committed:
Add support for creating local and asynchronous topics in `Topic.java`.

teosarca committed:
Introduce `PurchaseOrderProjectListener` and `Dispatcher` to propagate project information from purchase orders to related entities.

adi-stefan committed:
QA

teosarca committed:
Add tests and JSON serialization support for `ProjectCreatedEvent`; refine `OrderAndLineId` JSON handling.

- Added a unit test for `ProjectCreatedEvent` serialization/deserialization.
- Implemented `@Jacksonized` annotation for `ProjectCreatedEvent`.
- Improved `OrderAndLineId` with `@JsonCreator` and `@JsonValue` for proper JSON mapping.
- Addressed type mismatch in `purchaseOrderLineIds`.

teosarca committed:
Merge remote-tracking branch 'origin/new_dawn_uat_PO_ProjectId_To_SO' into new_dawn_uat_PO_ProjectId_To_SO

adi-stefan committed:
QA

adi-stefan committed:
Merge branch 'new_dawn_uat' into new_dawn_uat_PO_ProjectId_To_SO

adi-stefan committed:
Disable Sales/Purchase Order project type in vanilla metasfresh

adi-stefan committed:
Refactor project assignment logic in `C_Order_Project` to ensure consistent handling of null project IDs on order lines.

adi-stefan committed:
Remove redundant empty comments in `C_Order_Project`.

adi-stefan committed:
Update test snapshots to reflect OrderAndLineId serialization

adi-stefan committed:
Clarify documentation for `PurchaseOrderProjectListener` regarding project ID propagation.

adi-stefan committed:
Re-disable Sales/Purchase Order project type at the end of the cucumber.
